#### https://www.kaggle.com/primaryobjects/voicegender Çalışması

### features (öznitelikler)

meanfreq: mean frequency (in kHz) = ortalama frekans 

sd: standard deviation of frequency = frekansın standart sapması

median: median frequency (in kHz) = ortanca frekans

Q25: first quantile (in kHz) = ilk çeyrek

Q75: third quantile (in kHz) = üçüncü çeyrek

IQR: interquantile range (in kHz) = interquantile aralığı 

skew: skewness  = çarpıklık

kurt: kurtosis  =  kurtosis

sp.ent: spectral entropy = spektral entropi
  
sfm: spectral flatness = spektral düzlük

mode: mode frequency = mod frekansı

centroid: frequency centroid  = frekans centroid

peakf: peak frequency (frequency with highest energy) = tepe frekansı (en yüksek enerjili frekans)

meanfun: average of fundamental frequency measured across acoustic signal = akustik sinyal ile ölçülen temel frekansın ortalaması

minfun: minimum fundamental frequency measured across acoustic signal =  akustik sinyalde ölçülen minimum temel frekans

maxfun: maximum fundamental frequency measured across acoustic signal = akustik sinyal ile ölçülen maksimum temel frekans

meandom: average of dominant frequency measured across acoustic signal = akustik sinyal ile ölçülen baskın frekansın ortalaması

mindom: minimum of dominant frequency measured across acoustic signal = akustik sinyalde ölçülen minimum baskın frekans

maxdom: maximum of dominant frequency measured across acoustic signal = akustik sinyalde ölçülen maksimum baskın frekans
 
dfrange: range of dominant frequency measured across acoustic signal = akustik sinyalde ölçülen dominant frekans aralığı

modindx:

modulation index. Calculated as the accumulated absolute difference between adjacent measurements of fundamental 
frequencies divided by the frequency range 

modülasyon dizini. Temel bitişik ölçümler arasındaki birikmiş mutlak fark olarak hesaplanır
frekans aralığının frekans aralığına bölünmesiyle
 
### Target (Hedef)
label: male(0) or female(1) =  erkek(0) veya kadın(1) 

# Logistic Regression( clasification algorithm )

mini bir deep learning algoritmasıdır.

https://medium.com/@ODSC/logistic-regression-with-python-ede39f8573c7 (ingilizce)

https://www.udemy.com/course/machine-learning-ve-python-adan-zye-makine-ogrenmesi-4/ (ücretli kurs)

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html

In [2]:
# kütüphaneler
import pandas as pd
import numpy as np 

In [3]:
# veriyi okuma işlemi 
df = pd.read_csv('voice.csv')
df.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

In [4]:
# male ve female miktarı 
df['label'].value_counts()

female    1584
male      1584
Name: label, dtype: int64

In [5]:
# datamızın boyutu
df.shape

(3168, 21)

In [6]:
#datadaki kolonlar hakkında boş değer,kolon tipi ve total veriyi döndürür
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3168 entries, 0 to 3167
Data columns (total 21 columns):
meanfreq    3168 non-null float64
sd          3168 non-null float64
median      3168 non-null float64
Q25         3168 non-null float64
Q75         3168 non-null float64
IQR         3168 non-null float64
skew        3168 non-null float64
kurt        3168 non-null float64
sp.ent      3168 non-null float64
sfm         3168 non-null float64
mode        3168 non-null float64
centroid    3168 non-null float64
meanfun     3168 non-null float64
minfun      3168 non-null float64
maxfun      3168 non-null float64
meandom     3168 non-null float64
mindom      3168 non-null float64
maxdom      3168 non-null float64
dfrange     3168 non-null float64
modindx     3168 non-null float64
label       3168 non-null object
dtypes: float64(20), object(1)
memory usage: 519.9+ KB


In [7]:
#rastgele 3 veri 
df.sample(3)

meanfreq        sd    median       Q25       Q75       IQR      skew  \
2887  0.213309  0.041785  0.220000  0.198462  0.241154  0.042692  2.218113   
1850  0.206335  0.063949  0.226853  0.186820  0.248940  0.062120  2.063404   
1061  0.194565  0.057716  0.217628  0.137756  0.240513  0.102756  2.033257   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
2887  8.374674  0.864995  0.296586  ...  0.213309  0.158967  0.046921   
1850  8.360208  0.906087  0.435548  ...  0.206335  0.138612  0.016667   
1061  7.846657  0.907278  0.369606  ...  0.194565  0.124894  0.048290   

        maxfun   meandom    mindom    maxdom   dfrange   modindx   label  
2887  0.265193  1.930921  0.023438  6.375000  6.351562  0.183340  female  
1850  0.266667  0.183002  0.054688  0.695312  0.640625  0.107470  female  
1061  0.279070  1.907292  0.023438  8.484375  8.460938  0.128702    male  

[3 rows x 21 columns]

In [8]:
#featurelar arası bağlantı 
df.corr()

meanfreq        sd    median       Q25       Q75       IQR  \
meanfreq  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
sd       -0.739039  1.000000 -0.562603 -0.846931 -0.161076  0.874660   
median    0.925445 -0.562603  1.000000  0.774922  0.731849 -0.477352   
Q25       0.911416 -0.846931  0.774922  1.000000  0.477140 -0.874189   
Q75       0.740997 -0.161076  0.731849  0.477140  1.000000  0.009636   
IQR      -0.627605  0.874660 -0.477352 -0.874189  0.009636  1.000000   
skew     -0.322327  0.314597 -0.257407 -0.319475 -0.206339  0.249497   
kurt     -0.316036  0.346241 -0.243382 -0.350182 -0.148881  0.316185   
sp.ent   -0.601203  0.716620 -0.502005 -0.648126 -0.174905  0.640813   
sfm      -0.784332  0.838086 -0.661690 -0.766875 -0.378198  0.663601   
mode      0.687715 -0.529150  0.677433  0.591277  0.486857 -0.403764   
centroid  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
meanfun   0.460844 -0.466281  0.414909  0.545035  0.155091 -0.534462   
minfun    0.383937 -0.345609  0.337602  0.320994  0.258002 -0.222680   
maxfun    0.274004 -0.129662  0.251328  0.199841  0.285584 -0.069588   
meandom   0.536666 -0.482726  0.455943  0.467403  0.359181 -0.333362   
mindom    0.229261 -0.357667  0.191169  0.302255 -0.023750 -0.357037   
maxdom    0.519528 -0.482278  0.438919  0.459683  0.335114 -0.337877   
dfrange   0.515570 -0.475999  0.435621  0.454394  0.335648 -0.331563   
modindx  -0.216979  0.122660 -0.213298 -0.141377 -0.216475  0.041252   

              skew      kurt    sp.ent       sfm      mode  centroid  \
meanfreq -0.322327 -0.316036 -0.601203 -0.784332  0.687715  1.000000   
sd        0.314597  0.346241  0.716620  0.838086 -0.529150 -0.739039   
median   -0.257407 -0.243382 -0.502005 -0.661690  0.677433  0.925445   
Q25      -0.319475 -0.350182 -0.648126 -0.766875  0.591277  0.911416   
Q75      -0.206339 -0.148881 -0.174905 -0.378198  0.486857  0.740997   
IQR       0.249497  0.316185  0.640813  0.663601 -0.403764 -0.627605   
skew      1.000000  0.977020 -0.195459  0.079694 -0.434859 -0.322327   
kurt      0.977020  1.000000 -0.127644  0.109884 -0.406722 -0.316036   
sp.ent   -0.195459 -0.127644  1.000000  0.866411 -0.325298 -0.601203   
sfm       0.079694  0.109884  0.866411  1.000000 -0.485913 -0.784332   
mode     -0.434859 -0.406722 -0.325298 -0.485913  1.000000  0.687715   
centroid -0.322327 -0.316036 -0.601203 -0.784332  0.687715  1.000000   
meanfun  -0.167668 -0.194560 -0.513194 -0.421066  0.324771  0.460844   
minfun   -0.216954 -0.203201 -0.305826 -0.362100  0.385467  0.383937   
maxfun   -0.080861 -0.045667 -0.120738 -0.192369  0.172329  0.274004   
meandom  -0.336848 -0.303234 -0.293562 -0.428442  0.491479  0.536666   
mindom   -0.061608 -0.103313 -0.294869 -0.289593  0.198150  0.229261   
maxdom   -0.305651 -0.274500 -0.324253 -0.436649  0.477187  0.519528   
dfrange  -0.304640 -0.272729 -0.319054 -0.431580  0.473775  0.515570   
modindx  -0.169325 -0.205539  0.198074  0.211477 -0.182344 -0.216979   

           meanfun    minfun    maxfun   meandom    mindom    maxdom  \
meanfreq  0.460844  0.383937  0.274004  0.536666  0.229261  0.519528   
sd       -0.466281 -0.345609 -0.129662 -0.482726 -0.357667 -0.482278   
median    0.414909  0.337602  0.251328  0.455943  0.191169  0.438919   
Q25       0.545035  0.320994  0.199841  0.467403  0.302255  0.459683   
Q75       0.155091  0.258002  0.285584  0.359181 -0.023750  0.335114   
IQR      -0.534462 -0.222680 -0.069588 -0.333362 -0.357037 -0.337877   
skew     -0.167668 -0.216954 -0.080861 -0.336848 -0.061608 -0.305651   
kurt     -0.194560 -0.203201 -0.045667 -0.303234 -0.103313 -0.274500   
sp.ent   -0.513194 -0.305826 -0.120738 -0.293562 -0.294869 -0.324253   
sfm      -0.421066 -0.362100 -0.192369 -0.428442 -0.289593 -0.436649   
mode      0.324771  0.385467  0.172329  0.491479  0.198150  0.477187   
centroid  0.460844  0.383937  0.274004  0.536666  0.229261  0.519528   
meanfun   1.000000  0.339387  0.311950  0.270840

In [9]:
# değerleri aynı olan iki kolonumuz var datanın içinde 
df1 = pd.DataFrame({'centroid': df.centroid, 'meanfreq ': df.meanfreq})  
df1.tail()

centroid  meanfreq 
3163  0.131884   0.131884
3164  0.116221   0.116221
3165  0.142056   0.142056
3166  0.143659   0.143659
3167  0.165509   0.165509

In [10]:
# aynı değerlere sahip olan colonlardan biri silindi (drop centroid)
df = df.drop(columns='centroid') 

# datadaki kolonlar
print(list(df.columns))

['meanfreq', 'sd', 'median', 'Q25', 'Q75', 'IQR', 'skew', 'kurt', 'sp.ent', 'sfm', 'mode', 'meanfun', 'minfun', 'maxfun', 'meandom', 'mindom', 'maxdom', 'dfrange', 'modindx', 'label']


In [11]:
# male = 1 ,female = 0 1/0 formatına dönüştürme
df.label = [0 if each == "male" else 1  for each in df.label] 

In [12]:
# rastgele 3 label colonu yeni değerleri 
df.label.sample(3)

601     0
94      0
1013    0
Name: label, dtype: int64

In [13]:
# labels
y = df.label.values
y[500:505]

array([0, 0, 0, 0, 0], dtype=int64)

In [14]:
# features
x = df.drop(["label"],axis = 1)
print(list(x.columns))

['meanfreq', 'sd', 'median', 'Q25', 'Q75', 'IQR', 'skew', 'kurt', 'sp.ent', 'sfm', 'mode', 'meanfun', 'minfun', 'maxfun', 'meandom', 'mindom', 'maxdom', 'dfrange', 'modindx']


In [15]:
# normalization
from sklearn import preprocessing
# Scale the features / özellikleri ölçeklendirme
scaler = preprocessing.StandardScaler()

# separate features from label / etiketten ayrı özellikler
x_scaled = scaler.fit_transform(x[x.columns.values[:]]) #features / özellikler
x_scaled[0]

array([-4.04924806,  0.4273553 , -4.22490077, -2.57610164, -5.69360723,
       -0.21477826,  2.29330585,  1.76294635, -0.03908279,  0.4715753 ,
       -2.14121031, -1.81203825, -1.0979981 ,  0.56595854, -1.5642046 ,
       -0.70840431, -1.43142165, -1.41913712, -1.45477229])

In [22]:
# train_test split 
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split( x_scaled, y, test_size=0.2006, random_state = 42)

In [23]:
print(x_train.shape,x_test.shape)

(2532, 19) (636, 19)


In [17]:
say0 = 0 
say1 = 0
for each in y_train:
    if each == 0:
        say0 += 1 # male sayısı
    else:
        say1 +=1 # female sayısı
        
print("0 sayısı : ", say0 ," 1 sayısı : " ,say1 ,"total : ", (y_train.size))


0 sayısı :  1246  1 sayısı :  1286 total :  2532


In [18]:
# logistic regresyon( clasification algorithm gerçekleştirilir ve model eğitilir)
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression(random_state=0, solver='lbfgs',multi_class='multinomial').fit(x_train, y_train)

In [19]:
#  model yardımı ile y değerleri tahmin edilir 
y_pred = logreg.predict(x_test)
y_pred[1]

0

In [20]:
# logregin benim yerime tahmin  et ve  doğruluğu hesaplaması için kullandığımız fonksiyon 
logreg.score(x_test,y_test)

0.9811320754716981

In [21]:
# accuracy(doğruluk) fonksiyonunda tahmin edilir
import sklearn.metrics as metrics

# kullanılan metrik accuracy_score  popüler olan kategorik doğruluk metriği 
metrics.accuracy_score(y_test, y_pred)

0.9811320754716981

In [22]:
# f1 metriği sonuçlar aynı 
metrics.f1_score(y_test, y_pred)

0.9799331103678929

### BASELİNE :0.9811320754716981

# CART
### Decision Tree

titanik: 

                     cinsiyet erkek
                     |             |
               evet |               | hayır
                   |                 |
              yaş > 9,5              yaşıyor
            e|         |h
            ölü       yaşıyor  
            
sistemi ile oluşturulan bir ağaç ile seslerin  değerlerinin kıza mı yoksa erkeğemi ait olduğunu buluyoruz.

https://www.udemy.com/course/machine-learning-ve-python-adan-zye-makine-ogrenmesi-4/ (ücretli kurs)

https://medium.com/machine-learning-researcher/decision-tree-algorithm-in-machine-learning-248fb7de819e (ingilizce)



In [23]:
# tablo hakkında kısa bilgi 
df.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm      mode   meanfun    minfun    maxfun  \
0   274.402906  0.893369  0.491918  0.000000  0.084279  0.015702  0.275862   
1   634.613855  0.892193  0.513724  0.000000  0.107937  0.015826  0.250000   
2  1024.927705  0.846389  0.478905  0.000000  0.098706  0.015656  0.271186   
3     4.177296  0.963322  0.727232  0.083878  0.088965  0.017798  0.250000   
4     4.333713  0.971955  0.783568  0.104261  0.106398  0.016931  0.266667   

    meandom    mindom    maxdom   dfrange   modindx  label  
0  0.007812  0.007812  0.007812  0.000000  0.000000      0  
1  0.009014  0.007812  0.054688  0.046875  0.052632      0  
2  0.007990  0.007812  0.015625  0.007812  0.046512      0  
3  0.201497  0.007812  0.562500  0.554688  0.247119      0  
4  0.712812  0.007812  5.484375  5.476562  0.208274      0

In [24]:
# import  Decision Tree
from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier() 

In [25]:
# modeli eğitmek 
dt.fit(x_train, y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [26]:
dt.score(x_test,y_test)

0.9669811320754716

In [27]:
# y' yi tahmin etmek 
y_pred = dt.predict(x_test)
y_pred[1]

0

In [28]:
# accuracy_score doğruluğu
metrics.accuracy_score(y_test, y_pred)

0.9669811320754716

In [29]:
# f1_score doğruluğu
metrics.f1_score(y_test, y_pred)

0.9655172413793103

### BASELİNE :0.9811320754716981 > 0.9654088050314465

#### Logistic Regression > CART(Decision Tree)

# SVM(Support Vector Machine)

http://bilgisayarkavramlari.sadievrenseker.com/2008/12/01/svm-support-vector-machine-destekci-vektor-makinesi/

https://www.youtube.com/watch?v=Y6RRHw9uN9o (ingilizce)

https://www.udemy.com/course/machine-learning-ve-python-adan-zye-makine-ogrenmesi-4/ (ücretli kurs)


In [43]:
# model import

from sklearn.svm import SVC

svm = SVC(random_state = 1,kernel = 'rbf')

In [44]:
# model eğitme
svm.fit(x_train,y_train)

c:\users\pc\.virtualenvs\adresgezgini_interns-0cta1nx1\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
    kernel='rbf', max_iter=-1, probability=False, random_state=1,
    shrinking=True, tol=0.001, verbose=False)

In [45]:
# Test the model with default parameters
# Accuracies
print(f'Traınıng accuracy: {svm.score(x_train,y_train)} Testing accuracy: {svm.score(x_test,y_test)}')

Traınıng accuracy: 0.985781990521327 Testing accuracy: 0.9827044025157232


#### Perform CV

In [46]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score as accuracy

In [47]:
tuned_parameters = [{'kernel': ['rbf', 'linear'], 'gamma': ['auto_deprecated', 'scale'],
                     'C': np.linspace(1e-9, 1, 5)}]

In [48]:
score = 'accuracy'
clf = GridSearchCV(SVC(), tuned_parameters, cv=10, scoring=score)

clf.fit(x_train, y_train)
print(f'Traınıng accuracy: {clf.best_score_} Testing accuracy: {accuracy(y_test, clf.predict(x_test))}')

c:\users\pc\.virtualenvs\adresgezgini_interns-0cta1nx1\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
c:\users\pc\.virtualenvs\adresgezgini_interns-0cta1nx1\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
c:\users\pc\.virtualenvs\adresgezgini_interns-0cta1nx1\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", Futu

c:\users\pc\.virtualenvs\adresgezgini_interns-0cta1nx1\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
c:\users\pc\.virtualenvs\adresgezgini_interns-0cta1nx1\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)
c:\users\pc\.virtualenvs\adresgezgini_interns-0cta1nx1\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", Futu

Traınıng accuracy: 0.981042654028436 Testing accuracy: 0.9827044025157232


c:\users\pc\.virtualenvs\adresgezgini_interns-0cta1nx1\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


Maybe the results can be ımproved using **feature selectıon**  
Some known feature selection algorithms are:  
* Pricipal component analysis (PCA)
* Recurssive reature selection (RFE)

---